### PR TITLE
examples: Added observability example

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds an Observability example. It consists of the Hello World example with Observability module turned on.

RELEASE NOTES:
* examples: Add Observability example

### CODE REVIEW

**Comment by zasweq** [comment]
Vindhya mentioned add a read me to that.

**Comment by ejona86** [comment]
Note, the o11y defer won't actually be run as-is, it'd require more complexity to handle Ctrl-C:
```diff
diff --git a/examples/features/observability/server/main.go b/examples/features/observability/server/main.go
index dc346683..8a7aef3c 100644
--- a/examples/features/observability/server/main.go
+++ b/examples/features/observability/server/main.go
@@ -25,6 +25,8 @@ import (
        "fmt"
        "log"
        "net"
+       "os"
+       "os/signal"
        "time"
 
        "google.golang.org/grpc"
@@ -63,7 +65,13 @@ func main() {
        s := grpc.NewServer()
        pb.RegisterGreeterServer(s, &server{})
        log.Printf("server listening at %v", lis.Addr())
+       c := make(chan os.Signal, 1)
+       signal.Notify(c, os.Interrupt)
+       go func() {
+               <-c
+               lis.Close()
+       }()
        if err := s.Serve(lis); err != nil {
-               log.Fatalf("failed to serve: %v", err)
+               log.Printf("failed to serve: %v", err)
        }
 }
```

Dunno what we want to do about that. Java has similar problems (that are more annoying to address).

**Comment by zasweq** [comment]
Sorry, I don't know what you mean by "Note, the o11y defer won't actually be run as-is"

**Comment by zasweq** [comment]
So the server binary must be closed by ctrl c? Your new code catches this signal and triggers a lis.Close()? This means our examples have been wrong ever since public preview?

File: examples/features/observability/client/main.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+func main() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	err := observability.Start(ctx)
+	if err != nil {
+		log.Fatalf("observability.Start() failed: %v", err)
+	}
+	defer observability.End()
```
Comment (dfawley) [inline examples/features/observability/client/main.go:50]: Maybe add some comments here to help what we've added stand out.

File: examples/features/observability/server/main.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// SayHello implements helloworld.GreeterServer
+func (s *server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+	log.Printf("Received: %v", in.GetName())
+	return &pb.HelloReply{Message: "Hello " + in.GetName()}, nil
+}
+
+func main() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	err := observability.Start(ctx)
```
Comment (dfawley) [inline examples/features/observability/server/main.go:56]: Ditto

File: security/advancedtls/examples/go.mod [hunk 3]
```diff
@@ -3,19 +3,19 @@ module google.golang.org/grpc/security/advancedtls/examples
 go 1.17
 
 require (
-	google.golang.org/grpc v1.52.0
+	google.golang.org/grpc v1.53.0-dev.0.20230315171901-a1e657ce53ba
```
Comment (dfawley) [inline security/advancedtls/examples/go.mod:6]: 1.54.0?

File: security/advancedtls/examples/go.mod [hunk 4]
```diff
@@ -3,19 +3,19 @@ module google.golang.org/grpc/security/advancedtls/examples
...
 	golang.org/x/net v0.8.0 // indirect
 	golang.org/x/sys v0.6.0 // indirect
 	golang.org/x/text v0.8.0 // indirect
-	google.golang.org/genproto v0.0.0-20230110181048-76db0878b65f // indirect
-	google.golang.org/protobuf v1.28.1 // indirect
+	google.golang.org/genproto v0.0.0-20230125152338-dcaf20b6aeaa // indirect
+	google.golang.org/protobuf v1.29.1 // indirect
 )
 
 replace google.golang.org/grpc => ../../..
```
Comment (dfawley) [inline security/advancedtls/examples/go.mod:21]: Perhaps we should remove all of these?

File: examples/features/observability/client/main.go (hunk 1 above)
Comment (zasweq) [reply examples/features/observability/client/main.go:50]: Done.

File: examples/features/observability/server/main.go (hunk 2 above)
Comment (zasweq) [reply examples/features/observability/server/main.go:56]: Done.

**Comment by zasweq** [comment]
Added comment and gave it's own go.mod in pr: https://github.com/grpc/grpc-go/pull/6149. Closing this now.

**Pull request closed without merging**